### PR TITLE
DAOS-13224 build test: Build and test on EL9

### PR DESCRIPTION
**Pull request by brianjmurrell**
Remove production of the version stash.  It's not necessary any longer.

Add a Fixes commit pragma to allow a PR to fix more than one issue that
might be in the skip list.

Remove libfabric along with any daos packages prior to installing daos,
to ensure that the correct libfabric is installed.

Remove use of lsb_release as it's gone in EL9.  Use `/etc/os-release`
instead.

s-nail replaces mail[x] for sending e-mail on EL9.

Remove ci/functional/required_packages.sh and move all functional test
dependency specifications into RPMs as Requires:.

Add a client-tests-mpich subpackage for mpich test dependencies.

Remove the "Scan EL 8 RPMs" stage due to `Jenkinsfile` size limitations.

Fixes: SRE-1899 DAOS-13225

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
  * No code related to failed tests was modified
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by JohnMalmberg — APPROVED** [review]

File: src/tests/ftest/mpiio/llnl_mpi4py.py [hunk 1]
```diff
@@ -13,6 +16,25 @@ class LlnlMpi4py(MpiioTests):
...
+        """
+        test_repo = self.params.get(name, '/run/test_repo/')
+        for packages in site.getsitepackages():
+            test_path = os.path.join(packages, test_repo)
+            if os.path.exists(test_path):
+                return test_path
+
+        self.fail('No test repo found in python site-packages')
+
+        return None
```
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply src/tests/ftest/mpiio/llnl_mpi4py.py:36]: @daltonbohning Normally I'd be amenable to such a change, but I go on sabbatical next Monday and @ipoddubn would like this (whole EL9 build and test stack -- of which this is just a part of) to be landed before I leave.  So this needs to land soon enough to allow the myriad (dozen or so) of test tools and middle ware that need the `daos-devel` from this PR to be available in Artifactory for them all to be able to build to be landed.

**Review by daltonbohning — COMMENTED** [review]
Are the `src/rdb/raft` changes expected to be pushed? I'm not familiar with how submodules are updated in the origin

**Review by daltonbohning — COMMENTED** [review]
To be clear: this builds on EL9, but Functional on El9 is still not working?
Also: shouldn't the HW stages have ran?

**Comment by brianjmurrell** [comment]
> Are the `src/rdb/raft` changes expected to be pushed? I'm not familiar with how submodules are updated in the origin

Yes.

> Are the `src/rdb/raft` changes expected to be pushed? I'm not familiar with how submodules are updated in the origin

Yes.  They update the raft module to build for EL9.

**Comment by brianjmurrell** [comment]
> To be clear: this builds on EL9, but Functional on El9 is still not working?

As of a few minutes ago, this will not build or test on EL9 on PRs -- it acts the same way as Leap 15 does.  That means it will build and test on landings and timed runs.  Until the ticketed issue for EL9 is resolved, I have suggested to @phender that the failing tests should be skipped on EL9.

But ultimately we need this to land before the issue is resolved as the person resolving that will need to be able to do EL9 build+test runs.

> Also: shouldn't the HW stages have ran?

They don't run when the VM functional stages fail.  But we also don't do anything with EL9 on the hardware stages.

But I am also noticing that I have conflicts in the package version with master, so I will merge it and run this again.  It should skip the EL9 stages in that run and so the HW stages should also run.

**Review by daltonbohning — DISMISSED** [review]

**Review by daltonbohning — DISMISSED** [review]

**Review by JohnMalmberg — APPROVED** [review]

File: utils/githooks/pre-commit.d/20-flake.sh [hunk 2]
```diff
@@ -21,6 +21,11 @@ if ! command -v flake8 > /dev/null 2>&1; then
     exit 0
 fi
 
+if [ ! -f .flake8 ]; then
+    echo "  No .flake8, skipping flake checks"
+    exit 0
+fi
+
```
Comment (ashleypittman) [inline utils/githooks/pre-commit.d/20-flake.sh:28]: Is this change intentional?

File: utils/githooks/pre-commit.d/40-pylint.sh [hunk 3]
```diff
@@ -10,10 +10,13 @@ echo "Pylint:"
...
-    git diff "$TARGET"... --name-only | ./utils/cq/daos_pylint.py --files-from-stdin
+
+if [ -f utils/cq/daos_pylint.py ]; then
+    if [ "$TARGET" = "HEAD" ]; then
+            echo "  Checking against HEAD"
+            git diff HEAD --name-only | ./utils/cq/daos_pylint.py --files-from-stdin
+    else
+            echo "  Checking against branch ${TARGET}"
+            git diff "$TARGET"... --name-only | ./utils/cq/daos_pylint.py --files-from-stdin
+    fi
```
Comment (ashleypittman) [inline utils/githooks/pre-commit.d/40-pylint.sh:21]: Same comment for here.  On this branch the files tested for do exist.

File: site_scons/env_modules.py [hunk 4]
```diff
@@ -107,6 +108,7 @@ def _mpi_module(self, mpi):
                 self._module_func('unload', to_unload)
 
         for to_load in load:
+            print(f"Trying to load {to_load}")
```
Comment (ashleypittman) [inline site_scons/env_modules.py:111]: Debug left in.

File: utils/githooks/pre-commit.d/20-flake.sh (hunk 2 above)
Comment (brianjmurrell) [reply utils/githooks/pre-commit.d/20-flake.sh:28]: It is.  I don't recall the specific issue though.  I am trying to make the githooks more portable to more repositories and must have run into some issue with `.flake8` being missing.

File: utils/githooks/pre-commit.d/40-pylint.sh (hunk 3 above)
Comment (brianjmurrell) [reply utils/githooks/pre-commit.d/40-pylint.sh:21]: Again, it is likely due to trying to use the same githooks in other repos.

File: site_scons/env_modules.py [hunk 5]
```diff
@@ -56,6 +56,7 @@ def _module_func(self, command, *arguments):  # pylint: disable=no-self-use
 
         # pylint: disable=consider-using-with
         try:
+            print(f"Going to run {cmd}")
```
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply site_scons/env_modules.py:59]: No.  I left that in intentionally.  I run into issues with this whole code path enough times that I think it's helpful to have that here.

File: site_scons/env_modules.py (hunk 4 above)
Comment (brianjmurrell) [reply site_scons/env_modules.py:111]: Ditto.

File: utils/githooks/pre-commit.d/20-flake.sh (hunk 2 above)
Comment (ashleypittman) [reply utils/githooks/pre-commit.d/20-flake.sh:28]: Nice idea.

File: site_scons/env_modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply site_scons/env_modules.py:59]: Indeed.  I think that is good though.  It's providing some transparency about how CI is setting up the environment for MPI.  It helps users that want to replicate CI have more understanding of what they need to do.

**Comment by brianjmurrell** [comment]
Gatekeeper: Please DO NOT land this.  I have removed daos-gatekeeper from the reviewers as all of the test-tool and middle-ware RPM PRs will need to land also or there will be massive amounts of test failures on `master` landings and timed runs -- where the EL9 test stage will run (it's only disabled for PRs, like Leap 15 is currently).

Given the time left this week, I don't think I can get all of those other PRs landed in a single day to prevent those massive failures for the 4 weeks I will be out on sabbatical starting Monday.

**Comment by brianjmurrell** [comment]
@JohnMalmberg @daltonbohning I'm hoping I can get some quick and timely reviews on the master merge that I had to do this PR for simply `daos.spec` and `debian/changelog` conflict update.  @daos-stack/daos-gatekeeper willing, I am hoping to get this landed as early as possible tomorrow morning so that I can start landing the other side of the stack (test tools and middleware), hopefully all within the day so that there are not massive failures due to them being missing in tomorrow night's daily run which will include EL9 as a tested platform.

**Review by daltonbohning — APPROVED** [review]

**Review by JohnMalmberg — APPROVED** [review]

File: src/tests/ftest/process_core_files.py [hunk 6]
```diff
@@ -250,9 +258,9 @@ def install_debuginfo_packages(self):
                     dnf_args.extend(
                         ["--enablerepo=*-debuginfo", "--exclude", "nvml-debuginfo", "libpmemobj",
                          "python36", "openmpi3", "gcc"])
-                elif self.is_el() and self.distro_info.version == "8":
+                elif self.is_el() and int(self.distro_info.version) >= 8:
                     dnf_args.extend(
-                        ["--enablerepo=*-debuginfo", "libpmemobj", "python3", "openmpi", "gcc"])
+                        ["libpmemobj", "python3", "openmpi", "gcc"])
```
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply src/tests/ftest/process_core_files.py:263]: I think that was (always?) superfluous.  Or perhaps some previous bug(s) that were preventing installing the necessary debug packages has been fixed.  In any case, core file testing is one thing that I focused on in this PR for all platforms.

**Comment by brianjmurrell** [comment]
I don't seem to be able to reply to https://github.com/daos-stack/daos/pull/12176#discussion_r1258946552 directly, so will have to create a new comment here.

> This results in quite a bit of debug that seems to only add the order of possible commands from what ia already printed, e.g.
> ...
> 
> I think it should be removed.

I left it in for transparency in how to reproduce what CI does to get the modules loaded.  I can remove it if you think it's too much and too little value.
